DD15 - Insolvent Contract (#201)
NethermindEth/Forta-Agents: DYDX-Bots/Insolvent_Contract/README.md

# Insolvent Contract Bot

## Description

This bot detects when the contract is insolvent. Insolvency is refers to when `totalBorrowerDebtBalance` is greater than `totalActiveBalanceCurrentEpoch`. The checks for both variables in every block.

> 

Reviewer: ` The checks for both variables in every block.` Do you mean `the bot checks ..`?

## Supported Chains

- Ethereum

## Alerts

- DYDX-15
  - Fired when `totalBorrowerDebtBalance` is greater than `totalActiveBalanceCurrentEpoch`.
  - Severity is always set to "Info".
  - Type is always set to "Info".
  - Metadata includes:
    - `totalBorrowerDebtBalance`: The total debt balance owed by borrowers.
    - `totalActiveBalanceCurrentEpoch`: The current epoch's total active balance.

## Test Data

> Note: To test the following, follow the comments in `utils.ts`.

The bot behaviour can be verified with the following contracts on the Kovan ETH testnet:

[0x4aaAF7a1a829b2130AF2Ec7A4A4F5FEd27D8eD1F](https://kovan.etherscan.io/address/0x4aaAF7a1a829b2130AF2Ec7A4A4F5FEd27D8eD1F) - `TestImplemenation`.

[0xe9511Faa2B2ccE548A5999b4bC3772e6a0f1C14A](https://kovan.etherscan.io/address/0xe9511Faa2B2ccE548A5999b4bC3772e6a0f1C14A) - `TestProxy`.

> You can test the bot on the Kovan ETH testnet by running `npm start`. Also, you can use the setter functions in the `testProxy` contract to toggle both `totalBorrowerDebtBalance` and `totalActiveBalanceCurrentEpoch` to test them in real time. This is because the bot checks these values in every block.

Reviewer: I'm not sure why this part is needed in the note: `This is because the bot checks these values in every block.`

Reviewer: I think you can generate a `test block` where there's an insolvency case by running these setters.

Author: > I'm not sure why this part is needed in the note: This is because the bot checks these values in every block.

It was me trying to explain why you could simply run `npm start` and toggle the values in real time.

> I think you can generate a test block where there's an insolvency case by running these setters.

I tried that several times, but what kept happening, is that at some point the block would be "too old" (this is my assumption) and it wouldn't generate a finding anymore and would actually error out.  I thought may it had to do with the provided RPC endpoint, since the bot would work again if I simply toggled the values again.

# Insolvent Contract Bot

## Description

This bot detects when the contract is insolvent. Insolvency refers to when the difference between `totalBorrowerDebtBalance` and `totalActiveBalanceCurrentEpoch` is greater than a set static threshold. The bot checks for both variables in every block.

> > Note: To set the static threshold to a desired amount, update the value of `THRESHOLD_AMOUNT` in `utils.ts`. Currently, it is set to `0`, and thus will detect **any** difference between the two varibles.

Reviewer: ```suggestion
> Note: To set the static threshold to a desired amount, update the value of `THRESHOLD_AMOUNT` in `utils.ts`. Currently, it is set to `0`, and thus will detect **all** insolvency cases.
```
I suggest this to help the client understand this direct link between this threshold and the insolvency detection.
